File: benchmarks/runners/sentil_runner.jl
using Sentil
using Printf

const FORMULA = "always[0, 100](eventually[0, 10](x > 5))"

now_ms() = time_ns() / 1.0e6

function summarize(samples::Vector{Float64})
    s = sort(samples)
    n = length(s)
    m = sum(s) / n
    v = n > 1 ? sum((x - m)^2 for x in s) / (n - 1) : 0.0
    pct(q) = s[clamp(round(Int, (n - 1) * q) + 1, 1, n)]
    return (mean = m, std = sqrt(v), min = s[1], p50 = pct(0.50), p99 = pct(0.99))
end

function cpu_model()
    info = Sys.cpu_info()
    isempty(info) ? "unknown" : strip(info[1].model)
end

function peak_rss_bytes()
    for line in eachline("/proc/self/status")
        startswith(line, "VmHWM:") && return parse(Int, split(line)[2]) * 1024
    end
    return -1
end

function emit(benchmark, formula, question, size, robustness, t, runs)
    rss = peak_rss_bytes()
    rss_field = rss >= 0 ? string(rss) : "null"
    @printf("{\"tool\":\"sentil\",\"version\":\"1.0.0\",\"language\":\"julia\",\"benchmark\":\"%s\",", benchmark)
    @printf("\"formula\":\"%s\",\"question\":\"%s\",\"size\":%d,\"robustness\":%.17g,", formula, question, size, robustness)
    @printf("\"timing\":{\"mean_ms\":%.17g,\"std_ms\":%.17g,\"min_ms\":%.17g,\"p50_ms\":%.17g,\"p99_ms\":%.17g},",
            t.mean, t.std, t.min, t.p50, t.p99)
    @printf("\"peak_rss_bytes\":%s,\"runs\":%d,\"hardware\":{\"cpu\":\"%s\",\"cores\":%d}}\n",
            rss_field, runs, cpu_model(), Sys.CPU_THREADS)
end

oracle_trace(n) = Trace(collect(0.0:(n - 1)), "x", [15.0 * sin(i * 0.1) for i in 0:(n - 1)])

function scalability()
    for n in (1000, 10000, 100000, 1000000, 10000000)
        runs = n <= 100000 ? 30 : 5
        trace = oracle_trace(n)
        monitor = Monitor(FORMULA)

        full_rob = robustness_signal(monitor, trace)[1]
        samples = Float64[]
        for _ in 1:runs
            t0 = now_ms()
            robustness_signal(monitor, trace)
            push!(samples, now_ms() - t0)
        end
        emit("scalability/length", FORMULA, "full_signal", n, full_rob, summarize(samples), runs)

        mon_rob = robustness(monitor, trace)
        samples = Float64[]
        for _ in 1:runs
            t0 = now_ms()
            robustness(monitor, trace)
            push!(samples, now_ms() - t0)
        end
        emit("scalability/length", FORMULA, "monitoring", n, mon_rob, summarize(samples), runs)
    end
end

function streaming()
    monitor = OnlineMonitor(FORMULA)
    index = symbol_index(monitor, "x")
    n = 1000000
    latencies = Vector{Float64}(undef, n)
    packed = zeros(1)
    last = 0.0
    for i in 0:(n - 1)
        packed[index] = 15.0 * sin(i * 0.1)
        t0 = now_ms()
        verdict = update_packed!(monitor, Float64(i), packed)
        latencies[i + 1] = now_ms() - t0
        last = verdict.lower
    end
    emit("streaming", FORMULA, "monitoring", n, last, summarize(latencies), n)
end

function main()
    suite = isempty(ARGS) ? "" : ARGS[1]
    if suite == "scalability"
        scalability()
    elseif suite == "streaming"
        streaming()
    else
        println(stderr, "unknown suite `$suite`; use `scalability` or `streaming`")
        exit(1)
    end
end

main()

File: benchmarks/runners/sentil_runner.java
import io.github.sedislab.sentil.Monitor;
import io.github.sedislab.sentil.OnlineMonitor;
import io.github.sedislab.sentil.Robustness;
import io.github.sedislab.sentil.Trace;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;

class SentilRunner {
    static final String FORMULA = "always[0, 100](eventually[0, 10](x > 5))";

    static double nowMs() {
        return System.nanoTime() / 1.0e6;
    }

    static final class Summary {
        final double mean;
        final double std;
        final double min;
        final double p50;
        final double p99;

        Summary(double[] samples) {
            double[] s = samples.clone();
            Arrays.sort(s);
            int n = s.length;
            double sum = 0;
            for (double x : s) {
                sum += x;
            }
            mean = sum / n;
            double acc = 0;
            for (double x : s) {
                acc += (x - mean) * (x - mean);
            }
            std = n > 1 ? Math.sqrt(acc / (n - 1)) : 0.0;
            min = s[0];
            p50 = s[Math.min(Math.max((int) Math.round((n - 1) * 0.50), 0), n - 1)];
            p99 = s[Math.min(Math.max((int) Math.round((n - 1) * 0.99), 0), n - 1)];
        }
    }

    static String cpuModel() {
        try {
            for (String line : Files.readAllLines(Path.of("/proc/cpuinfo"))) {
                if (line.startsWith("model name")) {
                    return line.substring(line.indexOf(':') + 1).trim();
                }
            }
        } catch (IOException ignored) {
        }
        return "unknown";
    }

    static long peakRssBytes() {
        try {
            for (String line : Files.readAllLines(Path.of("/proc/self/status"))) {
                if (line.startsWith("VmHWM:")) {
                    return Long.parseLong(line.split("\\s+")[1]) * 1024;
                }
            }
        } catch (IOException | NumberFormatException ignored) {
        }
        return -1;
    }

    static void emit(String benchmark, String question, long size, double robustness, Summary t,
            long runs) {
        long rss = peakRssBytes();
        String rssField = rss >= 0 ? Long.toString(rss) : "null";
        StringBuilder out = new StringBuilder();
        out.append("{\"tool\":\"sentil\",\"version\":\"0.3.0\",\"language\":\"java\",\"benchmark\":\"")
                .append(benchmark).append("\",");
        out.append(String.format("\"formula\":\"%s\",\"question\":\"%s\",\"size\":%d,\"robustness\":%.17g,",
                FORMULA, question, size, robustness));
        out.append(String.format(
                "\"timing\":{\"mean_ms\":%.17g,\"std_ms\":%.17g,\"min_ms\":%.17g,\"p50_ms\":%.17g,\"p99_ms\":%.17g},",
                t.mean, t.std, t.min, t.p50, t.p99));
        out.append(String.format("\"peak_rss_bytes\":%s,\"runs\":%d,\"hardware\":{\"cpu\":\"%s\",\"cores\":%d}}",
                rssField, runs, cpuModel(), Runtime.getRuntime().availableProcessors()));
        System.out.println(out);
    }

    static Trace oracleTrace(int n) throws Exception {
        double[] times = new double[n];
        double[] x = new double[n];
        for (int i = 0; i < n; i++) {
            times[i] = i;
            x[i] = 15.0 * Math.sin(i * 0.1);
        }
        Trace trace = Trace.create(times);
        trace.addSignal("x", x);
        return trace;
    }

    static void scalability() throws Exception {
        for (int n : new int[] {1000, 10000, 100000, 1000000, 10000000}) {
            int runs = n <= 100000 ? 30 : 5;
            try (Trace trace = oracleTrace(n); Monitor monitor = Monitor.parse(FORMULA)) {
                double fullRob = monitor.robustnessSignal(trace)[0];
                double[] samples = new double[runs];
                for (int r = 0; r < runs; r++) {
                    double t0 = nowMs();
                    monitor.robustnessSignal(trace);
                    samples[r] = nowMs() - t0;
                }
                emit("scalability/length", "full_signal", n, fullRob, new Summary(samples), runs);

                double monRob = monitor.robustness(trace);
                samples = new double[runs];
                for (int r = 0; r < runs; r++) {
                    double t0 = nowMs();
                    monitor.robustness(trace);
                    samples[r] = nowMs() - t0;
                }
                emit("scalability/length", "monitoring", n, monRob, new Summary(samples), runs);
            }
        }
    }

    static void streaming() throws Exception {
        try (OnlineMonitor monitor = OnlineMonitor.create(FORMULA)) {
            int index = (int) monitor.symbolIndex("x").orElseThrow(IllegalStateException::new);
            int n = 1000000;
            double[] latencies = new double[n];
            double[] packed = new double[1];
            double last = 0.0;
            for (int i = 0; i < n; i++) {
                packed[index] = 15.0 * Math.sin(i * 0.1);
                double t0 = nowMs();
                Robustness verdict = monitor.updatePacked(i, packed);
                latencies[i] = nowMs() - t0;
                last = verdict.lower();
            }
            emit("streaming", "monitoring", n, last, new Summary(latencies), n);
        }
    }

    public static void main(String[] args) throws Exception {
        String suite = args.length == 0 ? "" : args[0];
        if (suite.equals("scalability")) {
            scalability();
        } else if (suite.equals("streaming")) {
            streaming();
        } else {
            System.err.println("unknown suite `" + suite + "`; use `scalability` or `streaming`");
            System.exit(1);
        }
    }
}